bugsnag-plugin-android-ndk/src/main/jni/featureflags.c

```diff
@@ -0,0 +1,132 @@
+#include <string.h>
+
+#include "bugsnag_ndk.h"
+#include "featureflags.h"
+#include "utils/string.h"
+
+/*
+ * Implementation notes:
+ *
+ * We store Feature Flags in a dynamically allocated array, sorted by the
+ * feature flag 'name' string, allowing us to binary-search the array for
+ * duplicates.
+ *
+ * This provides a reasonable compromise between speed and size, since the array
+ * is no larger than the number of feature flags (not counting malloc padding)
+ * and is unlikely to be modified often. It also keeps testing simple, as the
+ * keys will always be in a known order.
+ *
+ * We resize the array using 'realloc' and 'memmove' to try and keep the
+ * overhead reasonable.
+ */
+
+static int feature_flag_index(const bsg_environment *env, const char *name) {
+  // simple binary search for a feature-flag by name
+  int low = 0;
+  int high = env->feature_flag_count - 1;
+
+  while (low <= high) {
+    int mid = (low + high) >> 1;
+    int cmp = strcmp(env->feature_flags[mid].name, name);
+
+    if (cmp < 0) {
+      low = mid + 1;
+    } else if (cmp > 0) {
+      high = mid - 1;
+    } else {
+      return mid; // found it
+    }
+  }
+
+  return -(low + 1);
+}
+
+static void *grow_array_for_index(void *array, const size_t element_count,
+                                  const size_t element_size,
+                                  const unsigned int index) {
+  void *new_array = realloc(array, (element_count + 1) * element_size);
```

Comment:
This can return `NULL` if the allocation failed - it'd be good to handle this edge case